pkg/cmd/cli/secrets/unlink_secret_from_object.go

```
@@ -0,0 +1,114 @@
+			if err := o.Complete(f, args, nil); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+			if err := o.Validate(); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := o.UnlinkSecrets(); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+
+		},
+	}
+
+	return cmd
+}
+
+func (o UnlinkSecretOptions) UnlinkSecrets() error {
+	serviceaccount, err := o.GetServiceAccount()
+	if err != nil {
+		return err
+	}
+
+	if err = o.unlinkSecretsFromServiceAccount(serviceaccount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// unlinkSecretsFromServiceAccount detatches pull and mount secrets from the service account.
+func (o UnlinkSecretOptions) unlinkSecretsFromServiceAccount(serviceaccount *kapi.ServiceAccount) error {
+	// All of the requested secrets must be present in either the Mount or Pull secrets
+	// If any of them are not present, we'll return an error and push no changes.
+	rmSecrets, err := o.GetSecrets()
+	if err != nil {
+		return err
+	}
+	rmSecretNames := o.GetSecretNames(rmSecrets)
+
+	var found bool
+
+	for _, secretname := range rmSecretNames.List() {
+		found = false
+
+		// Check the mount secrets
+		for i := len(serviceaccount.Secrets) - 1; i >= 0; i-- {
+			if secretname == serviceaccount.Secrets[i].Name {
+				found = true
+
+				// Unlink this entry from Secrets
+				serviceaccount.Secrets = append(serviceaccount.Secrets[:i], serviceaccount.Secrets[i+1:]...)
+			}
+		}
+
+		// Check the image pull secrets
+		for i := len(serviceaccount.ImagePullSecrets) - 1; i >= 0; i-- {
+			if secretname == serviceaccount.ImagePullSecrets[i].Name {
+				found = true
+
+				// Unlink this entry from ImagePullSecrets
+				serviceaccount.ImagePullSecrets = append(serviceaccount.ImagePullSecrets[:i], serviceaccount.ImagePullSecrets[i+1:]...)
+			}
+		}
+
+		if !found {
+			// This secret didn't appear in either secret type
+			return fmt.Errorf("%s is not associated with service account %s",
```

@smarterclayton I had originally discussed this with @pweil- in person. The reason I did it that way was to be protective of users making mistakes. If someone mistyped the name of a secret and received no error back from the CLI, then they're going to assume all is well. Meanwhile, the real secret is still exposed to the container.

For some linkages, I think the silent failure is fine, but when dealing with things like passwords, there's probably a really good reason why the admin is opting to unlink them and I think we probably don't want to quietly leave them linked.

If you agree, I'll work on the `--force` option.